Reference for `conan.tools.meson` classes and functions (#2533)
conan-io/docs: reference/tools/meson/mesontoolchain.rst

.. _conan_tools_meson_mesontoolchain:

MesonToolchain
==============

The ``MesonToolchain`` is the toolchain generator for Meson and it can be used in the ``generate()`` method
as follows:

.. code:: python

    from conan import ConanFile
    from conan.tools.meson import MesonToolchain

    class App(ConanFile):
        settings = "os", "arch", "compiler", "build_type"
        requires = "hello/0.1"
        options = {"shared": [True, False]}
        default_options = {"shared": False}

        def generate(self):
            tc = MesonToolchain(self)
            tc.preprocessor_definitions["MYDEFINE"] = "MYDEF_VALUE"
            tc.generate()

.. important::

    The ``MesonToolchain`` only works with the ``PkgConfigDeps`` generator.

>     Please, do not use other generators, as they can have overlapping definitions that can conflict.

Author: I can rephrase it, but I think it's important to know. The examples are fine to know how it works, but I guess that you should point it out here as well.

>     Please, do not use other generators, as they can have overlapping definitions that can conflict.

Reviewer: Yes, I think a rephrase would be ok. Something saying, "if you have dependencies" is the missing piece.

>     The ``MesonToolchain`` only works with the ``PkgConfigDeps`` generator.
>     Please, do not use other generators, as they can have overlapping definitions that can conflict.

Reviewer 2: ```suggestion
    When your recipe has dependencies ``MesonToolchain`` only works with the ``PkgConfigDeps`` generator.
    Please, do not use other generators, as they can have overlapping definitions that can conflict.
```

Generated files
-----------------

> This generates the following files after a :command:`conan install` (or when building the package
> in the cache) with the information provided in the ``generate()`` method as well as information
> translated from the current ``settings``, ``conf``, etc.:

Reviewer 2: ```suggestion
The ``MesonToolchain`` generates the following files after a :command:`conan install` (or when building the package
in the cache) with the information provided in the ``generate()`` method as well as information
translated from the current ``settings``, ``conf``, etc.:
```

- ``tools.apple:sdk_path`` argument for SDK path in case of Apple cross-compilation. It is used as value
  of the flag ``-isysroot``.
- ``tools.android:ndk_path`` argument for NDK path in case of Android cross-compilation. It is used to get
  some binaries like ``c``, ``cpp`` and ``ar`` used in ``[binaries]`` section from *conan_meson_cross.ini*.
- ``tools.build:cxxflags`` list of extra C++ flags that is used by ``cpp_args``.
- ``tools.build:cflags`` list of extra of pure C flags that is used by ``c_args``.
- ``tools.build:sharedlinkflags`` list of extra linker flags that is used by ``c_link_args`` and ``cpp_link_args``.
- ``tools.build:exelinkflags`` list of extra linker flags that is used by ``c_link_args`` and ``cpp_link_args``.

Cross-building for Apple and Android
-------------------------------------

The ``MesonToolchain`` adds all the flags required to cross-compile for Apple (MacOS M1, iOS, etc.) and Android.

**Apple**

It adds link flags ``-arch XXX``, ``-isysroot [SDK_PATH]`` and the minimum deployment target flag, e.g., ``-mios-version-min=8.0``
to the ``MesonToolchain`` ``c_args``, ``c_link_args``, ``cpp_args``, and ``cpp_link_args`` attributes, given the

> Conan settings for any Apple OS (iOS, watchOS, etc.) and the ``tools.apple:sdk_path`` configuration value like it's showed

Reviewer: ```suggestion
Conan settings for any Apple OS (iOS, watchOS, etc.) and the ``tools.apple:sdk_path`` configuration value like it's shown
```

in this example of host profile:

.. code-block:: text
    :caption: **ios_host_profile**

    [settings]
    os = iOS
    os.version = 10.0
    os.sdk = iphoneos
    arch = armv8
    compiler = apple-clang
    compiler.version = 12.0
    compiler.libcxx = libc++

    [conf]
    tools.apple:sdk_path=/my/path/to/iPhoneOS.sdk

**Android**

It initializes the ``MesonToolchain`` ``c``, ``cpp``, and ``ar`` attributes, which are needed to cross-compile for Android, given the

> Conan settings for Android and the ``tools.android:ndk_path`` configuration value like it's showed

Reviewer: ```suggestion
Conan settings for Android and the ``tools.android:ndk_path`` configuration value like it's shown
```

        options = {"shared": [True, False]}
        default_options = {"shared": False}

        def generate(self):
            tc = MesonToolchain(self)
            tc.preprocessor_definitions["MYDEFINE"] = "MYDEF_VALUE"
            tc.generate()

.. important::

    The ``MesonToolchain`` only works with the ``PkgConfigDeps`` generator.
    Please, do not use other generators, as they can have overlapping definitions that can conflict.

Generated files
-----------------

This generates the following files after a :command:`conan install` (or when building the package
in the cache) with the information provided in the ``generate()`` method as well as information
translated from the current ``settings``, ``conf``, etc.:

-  *conan_meson_native.ini*: if doing a native build.
-  *conan_meson_cross.ini*: if doing a cross-build (:ref:`conan_tools_build`).

conan_meson_native.ini
+++++++++++++++++++++++++

This file contains the definitions of all the Meson properties related to the Conan options
and settings for the current package, platform, etc. This includes but is not limited to the following:

* Detection of ``default_library`` from Conan settings.

>     * Based on existance/value of a option named ``shared``.

Reviewer 3: ```suggestion
    * Based on existence/value of an option named ``shared``.
```

* Detection of ``buildtype`` from Conan settings.

* Definition of the C++ standard as necessary.

* The Visual Studio runtime (``b_vscrt``), obtained from Conan input settings.

conan_meson_cross.ini
++++++++++++++++++++++++

This file contains the same information as the previous *conan_meson_native.ini*,
but with additional information to describe host, target, and build machines (such as the processor architecture).

Check out the meson documentation for more details on native and cross files:

* `Machine files <https://mesonbuild.com/Machine-files.html>`_
* `Native environments <https://mesonbuild.com/Native-environments.html>`_
* `Cross compilation <https://mesonbuild.com/Cross-compilation.html>`_

Customization
---------------

Attributes
+++++++++++++

definitions
^^^^^^^^^^^^

This attribute allows defining Meson project options:

.. code:: python

    def generate(self):
        tc = MesonToolchain(self)
        tc.definitions["MYVAR"] = "MyValue"
        tc.generate()

This is translated to:

> - One project options definition for ``MYVAR`` in ``conan_meson_native.init`` or ``conan_meson_cross.ini`` file.

Reviewer 3: ```suggestion
- One project options definition for ``MYVAR`` in ``conan_meson_native.ini`` or ``conan_meson_cross.ini`` file.
```

preprocessor_definitions
^^^^^^^^^^^^^^^^^^^^^^^^

This attribute allows defining compiler preprocessor definitions, for multiple configurations (Debug, Release, etc).

.. code:: python

    def generate(self):
        tc = MesonToolchain(self)
        tc.preprocessor_definitions["MYDEF"] = "MyValue"
        tc.generate()

This is translated to:

- One preprocessor definition for ``MYDEF`` in ``conan_meson_native.ini`` or ``conan_meson_cross.ini`` file.

> .. note::

Author: My fault. Bad copy&paste from Conan 1.x

- ``tools.android:ndk_path`` argument for NDK path in case of Android cross-compilation. It is used to get
  some binaries like ``c``, ``cpp`` and ``ar`` used in ``[binaries]`` section from *conan_meson_cross.ini*.
- ``tools.build:cxxflags`` list of extra C++ flags that is used by ``cpp_args``.
- ``tools.build:cflags`` list of extra of pure C flags that is used by ``c_args``.
- ``tools.build:sharedlinkflags`` list of extra linker flags that is used by ``c_link_args`` and ``cpp_link_args``.
- ``tools.build:exelinkflags`` list of extra linker flags that is used by ``c_link_args`` and ``cpp_link_args``.

Cross-building for Apple and Android
-------------------------------------

The ``MesonToolchain`` adds all the flags needed to cross-compile for Apple (MacOS M1, iOS, etc.) and Android.

**Apple**

It adds link flags ``-arch XXX``, ``-isysroot [SDK_PATH]`` and the minimum deployment target flag, e.g., ``-mios-version-min=8.0``
to the ``MesonToolchain`` ``c_args``, ``c_link_args``, ``cpp_args`` and ``cpp_link_args`` attributes.

**Android**

It initializes the ``MesonToolchain`` ``c``, ``cpp``, and ``ar`` attributes, which are needed to cross-compile for Android. For instance:

> * ``c == $TOOLCHAIN/bin/llvm-ar``

Author: ok, let me rephrase it a little bit

- ``tools.build:cxxflags`` list of extra C++ flags that will be used by ``cpp_args``.
- ``tools.build:cflags`` list of extra of pure C flags that will be used by ``c_args``.
- ``tools.build:sharedlinkflags`` list of extra linker flags that will be used by ``c_link_args`` and ``cpp_link_args``.
- ``tools.build:exelinkflags`` list of extra linker flags that will be used by ``c_link_args`` and ``cpp_link_args``.

Extending and advanced customization
------------------------------------

Using the toolchain in developer flow
+++++++++++++++++++++++++++++++++++++

One of the advantages of using Conan toolchains is that they can help to achieve the exact same build
with local development flows, than when the package is created in the cache.

With the ``MesonToolchain`` it is possible to do:

.. code:: bash

    # Lets start in the folder containing the conanfile.py
    $ mkdir build && cd build
    # Install both debug and release deps and create the toolchain
    $ conan install ..
    # the build type Release is encoded in the toolchain already.
    # This conan_meson_native.iniis specific for release
    $ meson setup --native-file conan_meson_native.ini build .

>     $ meson compile -C build

Reviewer 3: Agree

Cross-building for Apple and Android
+++++++++++++++++++++++++++++++++++++

It deserves a special mention because ``MesonToolchain`` is automatically adding all the flags needed
to cross-compile for Apple (MacOS M1, iOS, etc.) and Android.

**Apple**

It'll add link flags like ``-arch XXX``, ``-isysroot [SDK_PATH]`` and the minimum deployment target flag, e.g., ``-mios-version-min=8.0``
into Meson ``c_args``, ``c_link_args``, ``cpp_args`` and ``cpp_link_args`` built-in options.

**Android**

It'll initialize the ``c``, ``cpp`` and ``ar`` variables which are needed to cross-compile for Android. For instance:

* ``c == $TOOLCHAIN/bin/llvm-ar``
* ``cpp == $TOOLCHAIN/bin/$TARGET$API-clang``
* ``ar == $TOOLCHAIN/bin/$TARGET$API-clang++``

Where:

* ``$TOOLCHAIN``: ``[NDK_PATH]/toolchains/llvm/prebuilt/[OS_BUILD]-x86_64/bin``.
* ``$TARGET``: target triple, e.g., for ``armv8`` will be ``aarch64-linux-android``.
* ``$API``: Android API version.

Besides that, you'll always be able to change any of these variables before being applied thanks

> to the ``MesonToolchain`` class interface. For instance:

Reviewer: Which variables? only the ones for cross-building or all of the documented in the class below? Not sure if it is clear enough.
